### PR TITLE
Don't use flock on NFS mounts

### LINKED ISSUE #2615: cargo update hangs forever
I update to latest nightly and I have issues building my project. It seems `cargo update` hangs when updating `Cargo.lock`. Here is the output of strace after I manually removed the existing `Cargo.lock`:

```
...
madvise(0x7f4e9cb73000, 28672, MADV_DONTNEED) = 0
open("/cs/max/tools/test/Cargo.lock", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/cs/max/tools/test/Cargo.lock", O_RDWR|O_CREAT|O_CLOEXEC, 0666) = 3
ioctl(3, FIOCLEX)                       = 0
flock(3, LOCK_EX|LOCK_NB 
```

In hangs forever after the `flock` call.

Versions:
```
cargo 0.11.0-nightly (867627c 2016-04-21)
rustc 1.10.0-nightly (b5ba5923f 2016-04-21)
```

Details about the system where I am trying this:

```
Linux <removed> 2.6.32-358.el6.x86_64 #1 SMP Tue Jan 29 11:47:41 EST 2013 x86_64 x86_64 x86_64 GNU/Linux

# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 6.4 (Santiago)
```

### PR DESCRIPTION
**Pull request by alexcrichton**
Completely skip file locking when we detect an NFS mount via `statfs`.

Closes #2615

### CODE REVIEW

File: src/cargo/util/flock.rs [hunk 1]
```diff
@@ -263,9 +278,34 @@ fn acquire(config: &Config,
...
+        let path = match CString::new(path.as_os_str().as_bytes()) {
+            Ok(path) => path,
+            Err(_) => return false,
+        };
+
+        unsafe {
+            let mut buf: libc::statfs = mem::zeroed();
+            let r = libc::statfs(path.as_ptr(), &mut buf);
+
+            r == 0 && buf.f_type == libc::NFS_SUPER_MAGIC
```
Comment (brson) [inline src/cargo/util/flock.rs]: After googling this I'm even more curious. Just magic stuff.

**Comment by brson** [comment]
Do you know what happens when using NFS on other Oses?

**Comment by alexcrichton** [comment]
I have no idea :(

**Comment by Kerosene2000** [comment]
Here is an examples of [lock file error on an nfs file system on OS X](https://github.com/Manishearth/rust-clippy/issues/943).

**Comment by alexcrichton** [comment]
@Kerosene2000 thanks for the report! I'll send a PR soon

**Comment by alexcrichton** [comment]
(PR is https://github.com/rust-lang/cargo/pull/2720)

**Pull request merged**